File: lib/domain/entities/application.dart
import 'package:equatable/equatable.dart';

/// The [Application] class is a **application**, **app** or **package** installed
/// in client device.
class Application extends Equatable {
  const Application({
    required this.packageName,
    required this.category,
    required this.enabled,
  });

  final String packageName;

  final Category category;

  final bool enabled;

  @override
  List<Object?> get props => [packageName, category, enabled];
}

/// This [enumeration] is possibles categories of an application
enum Category {
  accessibility,
  audio,
  game,
  image,
  maps,
  news,
  productivity,
  social,
  undefined,
}

File: android/src/main/kotlin/com/leonardo/have_app/domain/entities/Application.kt
package com.leonardo.have_app.domain.entities

import kotlinx.serialization.Serializable

/**
 * A application found in the system
 *
 * @property packageName is name of this application
 * @property category is category of this application
 * @property enabled when false, indicates that all components within this application are
 * considered disabled.
 *
 */
@Serializable
open class Application(
    private val packageName: String,
    private val category: Category,
    private val enabled: Boolean,
) {
    override fun equals(other: Any?): Boolean {
        if (other is Application) {
            return (
                packageName == other.packageName &&
                category == other.category &&
                enabled == other.enabled
            )
        }

        return super.equals(other)
    }
}


@Serializable
enum class Category(val value: Int) {
    ACCESSIBILITY(8),
    AUDIO(1),
    GAME(0),
    IMAGE(3),
    MAPS(6),
    NEWS(5),
    PRODUCTIVITY(7),
    SOCIAL(4),
    UNDEFINED(-1),
}

fun getByValue(value: Int): Category {
    return Category.values().first { it.value == value }
}